src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -155,18 +142,14 @@ private static void checkpointRestore1(long jcmdStream) throws
         if (retCode != JVM_CHECKPOINT_OK) {
             if (checkpointException == null) {
                 checkpointException = new CheckpointException();
+                if (messages.length == 0) {
+                    checkpointException.addSuppressed(new RuntimeException("Native checkpoint failed"));
+                }
```

Comment:
codes.length == messages.length, and they are used to communicate native descriptors problems (handled in translateJVMExceptions). So a problem should be reported based on the retCode below, if that is not reported already (?)

Reply:
When the checkpoint does not fail in JVM code (checking open FDs) but the engine returns non-zero status, the retCode is `JVM_CHECKPOINT_ERROR` but the messages are empty. In this case the exception would look just like CheckpointException without any explanation nor stack trace, and the user would be rather clueless.
We could add to this message a hint to check out `dump4.log`, but I am not entirely sure about this as it breaks encapsulation of CR engine (alternative implementations... though we might make CRIU use 'standardized' location) and it might not exist at all, e.g. when the CRIU binary does not exist at all.

Reply:
Regarding providing more details, in general, that should be handled by the engine. The criuengine (as the entity controlling the dump4.log name and the path) has a chance to print a few lines from the log. E.g. by calling `system("tail ...")`, although not very pretty, for that will be better than nothing.